### PR TITLE
Remove rank 0 restrictions from logger

### LINKED ISSUE #8589: Remove rank restrictions on loggers
## 🚀 Feature

Allow loggers to log on non rank 0 processes.

### Motivation

We want to log system data across all ranks (e.g.  using a custom logger to log GPU memory, latency stats, etc). This is important if we want to detect stragglers or anomalous conditions in a rank.

### Pitch

The trainer should not check that the process is on global rank 0 before calling the logger.

We already have rank_zero_experiment (https://github.com/PyTorchLightning/pytorch-lightning/blob/a64cc373946a755ce6c3aef57c1be607dfe29a0c/pytorch_lightning/loggers/base.py#L33) and @rank_zero_only decorating all loggers. We can simply remove the global zero checks in logger_connector.py here (https://github.com/PyTorchLightning/pytorch-lightning/blob/6b47cbe3ca8aa3fd82211bc9fa32e734753a6950/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L105-L109), and here (https://github.com/PyTorchLightning/pytorch-lightning/blob/6b47cbe3ca8aa3fd82211bc9fa32e734753a6950/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L180).

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

Fixes #8589

This PR removes is_global_zero check before logging in the Trainer. It does not change the rank-zero-only behavior that the logger implementations already have and instead enables implementations of loggers that do need to log on each rank when designed that way.

### Does your PR introduce any breaking changes? If yes, please list them.

All existing loggers already have guards to only run on rank 0.

But any logger implemented by users who rely on this might fail now.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

EDIT(@awaelchli) updated description to motivate these changes

### CODE REVIEW

File: tests/loggers/test_all.py [hunk 1]
```diff
@@ -343,6 +343,47 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
...
+    def name(self):
+        return "AllRank"
+
+    def log_hyperparams(self, _params, *_args, **_kwargs):
+        pass
+
+
+class AllRanksLoggerCheck(Callback):
+    def on_train_batch_start(self, trainer, _pl_module, _batch, _batch_idx, _dataloader_idx):
+        assert trainer.logger.experiment() == AllRankLogger.exp
```
Comment (ananthsub) [inline tests/loggers/test_all.py:367]: please add typehints even for the test given the overall push to strongly type the codebase

File: tests/loggers/test_all.py [hunk 2]
```diff
@@ -343,6 +343,47 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
...
+
+
+@RunIf(skip_windows=True)
+def test_logger_created_on_all_ranks(tmpdir):
+    logger = AllRankLogger()
+    model = BoringModel()
```
Comment (ananthsub) [inline tests/loggers/test_all.py:373]: does this check that data from each rank is actually available in the logger? 

- you could extend the boring model and log the rank in the training step
- then in the all rank logger, you can locally store the logged metrics
- then the callback on fit end can then assert that the trainer's rank maps to the value inside of the logger to confirm that this does run on all ranks

File: tests/loggers/test_all.py [hunk 3]
```diff
@@ -343,6 +343,47 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
...
+    trainer = Trainer(
+        logger=logger,
+        default_root_dir=tmpdir,
+        accelerator="ddp_cpu",
+        num_processes=2,
+        max_steps=1,
+        checkpoint_callback=True,
```
Comment (ananthsub) [inline tests/loggers/test_all.py:380]: does the checkpoint callback need to be enabled? it will make the test run slower

File: CHANGELOG.md [hunk 4]
```diff
@@ -152,6 +152,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Improved error messages in `replace_sampler` when the `DataLoader` attributes are not included in the signature or the signature is missing optional arguments ([#8519](https://github.com/PyTorchLightning/pytorch-lightning/pull/8519))
 - Moved `DeviceDtypeModuleMixin` and `HyperparametersMixin` mixin to `core` ([#8396](https://github.com/PyTorchLightning/pytorch-lightning/pull/8396))
 - Return the `default_root_dir` as the `log_dir` when the logger is a `LoggerCollection` ([#8187](https://github.com/PyTorchLightning/pytorch-lightning/pull/8187))
+- Remove restriction that loggers can only log on rank 0 ([#8608](https://github.com/PyTorchLightning/pytorch-lightning/pull/8608))
```
Comment (ananthsub) [inline CHANGELOG.md:155]: this should go into a new section of the changelog for v1.5

@carmocca @awaelchli this would be a breaking change in case users implemented a custom logger and didn't explicitly check for rank zero only. i don't see a way to get around this though

File: tests/loggers/test_all.py [hunk 5]
```diff
@@ -343,6 +343,47 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
     assert trainer.state.finished, f"Training failed with {trainer.state}"
 
 
+class AllRankLogger(LightningLoggerBase):
```
Comment (ananthsub) [inline tests/loggers/test_all.py:346]: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/logging_/test_distributed_logging.py - the PR needs to update these tests. we could directly update them and remove the changes from here to confirm distributed logging works as expected.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 6]
```diff
@@ -103,9 +103,8 @@ def log_metrics(self, metrics: Dict[str, _METRIC], step: Optional[int] = None) -
             step = self.trainer.global_step
 
         # log actual metrics
-        if self.trainer.is_global_zero:
-            self.trainer.logger.agg_and_log_metrics(scalar_metrics, step=step)
-            self.trainer.logger.save()
+        self.trainer.logger.agg_and_log_metrics(scalar_metrics, step=step)
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:106]: Mind sharing the reasons why we would want all processes to log ?

File: CHANGELOG.md (hunk 4 above)
Comment (awaelchli) [reply CHANGELOG.md:155]: true, so this should definitely be in the changed section and I would also mention or make clear that the behavior for existing built-in loggers is unchanged. they still log only rank 0.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:106]: @tchaton I updated the PR description. The motivation here is to enable implementations of loggers that actually can lon every rank. The existing loggers we have internally already do have checks to log only on rank 0.

File: tests/loggers/test_all.py [hunk 7]
```diff
@@ -343,6 +343,47 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
...
+    exp = "test exp"
+
+    def experiment(self):
```
Comment (tchaton) [inline tests/loggers/test_all.py:349]: Add doctoring.

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,94 +4,70 @@ All notable changes to this project will be documented in this file.
...
-
-
 - Saved checkpoints will no longer use the type of a `Callback` as the key to avoid issues with unpickling ([#6886](https://github.com/PyTorchLightning/pytorch-lightning/pull/6886))
 
-
 - Improved string conversion for `ResultCollection` ([#8622](https://github.com/PyTorchLightning/pytorch-lightning/pull/8622))
 
-
--
+- Removed restrictions that logger can only log from rank 0. Existing logger behavior has not changed. ([#8608](https://github.com/PyTorchLightning/pytorch-lightning/pull/8608))
```
Comment (ananthsub) [inline CHANGELOG.md:29]:
```suggestion
- Removed restrictions in the trainer that loggers can only log from rank 0. Existing logger behavior has not changed. ([#8608](https://github.com/PyTorchLightning/pytorch-lightning/pull/8608))
```

**Review by ananthsub — APPROVED** [review]
LGTM! 

Unrelated to the PR, @carmocca @Borda is pre-commit expected to run on the changelog? is there a concern that this will lose git blame for the changelog?

**Comment by carmocca** [comment]
> is pre-commit expected to run on the changelog?

There isn't any bot merged to do it... what caused all those changes?

> is there a concern that this will lose git blame for the changelog?

Why would blame be important for the CHANGELOG? Each line links to the PR that made the change

**Comment by edward-io** [comment]
@carmocca sorry, prettify automatically formatted the changelog file. I've fixed this.

File: tests/trainer/logging_/test_distributed_logging.py [hunk 9]
```diff
@@ -45,12 +72,14 @@ def test_global_zero_only_logging_ddp_cpu(tmpdir):
...
         weights_summary=None,
+        logger=all_rank_logger,
+        log_every_n_steps=1,
     )
     trainer.fit(model)
 
 
 @RunIf(min_gpus=2)
-def test_global_zero_only_logging_ddp_spawn(tmpdir):
+def test_all_rank_logging_ddp_spawn(tmpdir):
```
Comment (ananthsub) [inline tests/trainer/logging_/test_distributed_logging.py:82]: @edward-io  this test is failing. by default, the trainer uses a tensorboard logger, which defines its `experiment` only on rank 0. so the assertion on L57 in the TestModel is failing because `self.logger.logs` is not populated with that logger.

**Review by carmocca — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: tests/trainer/logging_/test_distributed_logging.py [hunk 10]
```diff
@@ -12,31 +12,61 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import os
+from typing import Any, Dict, Optional, Union
 from unittest import mock
 from unittest.mock import Mock
 
 from pytorch_lightning import Callback, Trainer
+from pytorch_lightning.loggers import TensorBoardLogger
+from pytorch_lightning.loggers.base import LightningLoggerBase
```
Comment (ananthsub) [inline tests/trainer/logging_/test_distributed_logging.py:21]:
```suggestion
from typing import Any, Dict, Optional, Union
from unittest.mock import Mock

from pytorch_lightning import Callback, Trainer
from pytorch_lightning.loggers.base import LightningLoggerBase
```

**Review by justusschock — APPROVED** [review]

File: tests/trainer/logging_/test_distributed_logging.py [hunk 11]
```diff
@@ -12,31 +12,59 @@
...
+    Logs are saved to local variable `logs`.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.logs = {}
+        self.exp = object()
+
+    def experiment(self) -> Any:
+        return self.exp
```
Comment (ananthsub) [inline tests/trainer/logging_/test_distributed_logging.py:36]:
```suggestion
    @property
    def experiment(self) -> Any:
        return self.exp
```

File: tests/trainer/logging_/test_distributed_logging.py [hunk 12]
```diff
@@ -12,31 +12,59 @@
...
+
+    def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None):
+        self.logs.update(metrics)
+
+    def version(self) -> Union[int, str]:
+        return 1
```
Comment (ananthsub) [inline tests/trainer/logging_/test_distributed_logging.py:42]:
```suggestion
    @property
    def version(self) -> Union[int, str]:
        return 1
```

File: tests/trainer/logging_/test_distributed_logging.py [hunk 13]
```diff
@@ -12,31 +12,59 @@
...
+
+    def name(self) -> str:
+        return "AllRank"
```
Comment (ananthsub) [inline tests/trainer/logging_/test_distributed_logging.py:45]:
```suggestion
    @property
    def name(self) -> str:
        return "AllRank"
```

File: tests/trainer/logging_/test_distributed_logging.py [hunk 14]
```diff
@@ -23,23 +23,53 @@
 from tests.helpers.runif import RunIf
 
 
+class AllRankLogger(LightningLoggerBase):
```
Comment (awaelchli) [inline tests/trainer/logging_/test_distributed_logging.py:26]: above here, need to remove the unused import. 

I couldn't do it due to permission error pushing to your branch.
